Support pedestrian movement in Newtonian simulator

lola831 opened the pull request:
### Description
Enable pedestrian movement in Newtonian simulator by storing heading & speed in control and applying via setVelocity in step().

### Issue Link
[#355 ](https://github.com/BerkeleyLearnVerify/Scenic/issues/355)

### Checklist
- [x] I have tested the changes locally via `pytest` and/or other means
- [ ] I have added or updated relevant documentation
- [x] I have autoformatted the code with black and isort
- [ ] I have added test cases (if applicable)

### Additional Notes
N/A

dfremont commented on src/scenic/simulators/newtonian/simulator.py:
```diff
@@ -196,7 +196,22 @@ def isOnScreen(self, x, y):
+                    if obj.control["speed"] is not None
+                    else obj.speed
+                )
+                vel = Vector(0, s).rotatedBy(h)
+                obj.setVelocity(vel)
```
```suggestion
                obj.velocity = Vector(0, s).rotatedBy(h)
```
I think this is easier to understand, since you don't need to go look up what `setVelocity` does.

dfremont commented on tests/simulators/newtonian/test_newtonian.py:
```diff
@@ -59,3 +59,13 @@ def test_pickle(loadLocalScenario):
+    simulation = simulator.simulate(scene, maxSteps=8)
+    init = simulation.result.records["InitialPos"]
+    fin = simulation.result.records["FinalPos"]
+    assert init != fin, f"Pedestrian did not move."
```
```suggestion
    assert init != fin, "Pedestrian did not move."
```

dfremont commented on src/scenic/simulators/newtonian/simulator.py:
```diff
@@ -196,7 +196,22 @@ def isOnScreen(self, x, y):
+                )
+                s = (
+                    obj.control["speed"]
+                    if obj.control["speed"] is not None
```
If the `heading` and `speed` controls are both `None`, this will change the pedestrian to be moving in the direction `obj.heading`, even if it wasn't before (e.g. if you manually set the velocity so the pedestrian is moving in some other direction: `new Pedestrian with velocity (1, 1)`). If the pedestrian doesn't take any walking actions, we should leave its velocity unchanged.

dfremont reviewed (changes requested):
A couple minor and one more serious comment above. Also in testing this I came across #339 again; we should discuss how to fix that.

lola831 commented:
@dfremont  I’ve applied all your suggestions in commit 369035b, thanks!

lola831 commented:
I’ve updated the step function so pedestrians no longer get misclassified. I also added a test (test_pedestrian_velocity_vector) to check that a pedestrian with velocity (1, 1) moves northeast as expected.

dfremont commented on tests/simulators/newtonian/test_newtonian.py:
```diff
@@ -59,3 +59,27 @@ def test_pickle(loadLocalScenario):
+
+
+def test_pedestrian_movement(loadLocalScenario):
+    scenario = loadLocalScenario("pedestrian.scenic", mode2D=True)
```
For new tests, let's inline the Scenic code into the function unless there's a specific reason why the code needs to be in another file. It's easier to understand what the test does, and you can use the `getAssetPath` fixture to get the map path in a more robust way (like in the CARLA tests).

dfremont commented on tests/simulators/newtonian/test_newtonian.py:
```diff
@@ -59,3 +59,27 @@ def test_pickle(loadLocalScenario):
+    simulation = simulator.simulate(scene, maxSteps=8)
+    init = simulation.result.records["InitialPos"]
+    fin = simulation.result.records["FinalPos"]
+    assert init != fin, "Pedestrian did not move."
```
Let's robustify this a bit so that the condition wouldn't just be satisfied by roundoff error, etc. How about something like `init.distanceTo(fin) > 0.1`?

dfremont reviewed (changes requested):
It's working now -- just a couple minor nitpicks.

lola831 commented on src/scenic/simulators/newtonian/simulator.py:
```diff
@@ -35,7 +35,7 @@
 HEIGHT = 800
 MAX_ACCELERATION = 5.6  # in m/s2, seems to be a pretty reasonable value
-MAX_BRAKING = 4.6
+MAX_BRAKING = 6.8
```
I ran into this when testing pedestrianJaywalking.scenic — with MAX_BRAKING = 4.6, the car doesn’t stop in time and just drives past the pedestrian. Increasing it to 6.8 made the braking more realistic and let the ego stop as expected, similar to how it behaves in the MetaDrive simulator.

dfremont replied:
OK, thanks.

lola831 commented:
@dfremont Thanks for the feedback. I’ve applied your suggestions.

lola831 commented on tests/simulators/newtonian/test_newtonian.py:
```diff
@@ -6,7 +6,8 @@
 from scenic.domains.driving.roads import Network
 from scenic.simulators.newtonian import NewtonianSimulator
-from tests.utils import pickle_test, sampleScene, tryPickling
+from tests.conftest import getAssetPath
```
@dfremont Thanks! I've gone ahead and removed the `getAssetPath` import.

dfremont reviewed (approved):
Thanks, Lola!
